### DISCUSSION TITLE
Seeeduino XIAO  vs Teensy

### DISCUSSION THREAD

**Discussion by MarcosGogo** [tttapa/Control-Surface#501] [General]
Hey guys.
Has anyone used a [Seeeduino XIAO](https://www.seeedstudio.com/Seeeduino-XIAO-Arduino-Microcontroller-SAMD21-Cortex-M0+-p-4426.html) with this library? According to the [MIDI OVER USB](https://tttapa.github.io/Control-Surface-doc/Doxygen/d8/d4a/md_pages_MIDI-over-USB.html) documents, SAMD21 supports MIDI over USB natively. Could someone confirm this? If possible, what would be the advantages of a Teensy board over a Xiao?

**Comment by tttapa** [comment]
> According to the MIDI OVER USB documents, SAMD21 supports MIDI over USB natively.

This is correct, the chip supports it, but I've only ever considered the official Arduino Core and MIDIUSB libraries, I cannot guarantee that it'll work with a 3rd-party Core like Seeed's, although it will probably work. A first step is getting it to compile and to verify that Control Surface correctly selects MIDIUSB as the MIDI over USB backend for the Seeeduino board (this is done [here](https://github.com/tttapa/Control-Surface/blob/a061a3a0f021059bfd53c89c928271476714be7b/src/MIDI_Interfaces/USBMIDI/USBMIDI.hpp#L61)). All of this can be verified without actually buying a board.

I have never used any Seeeduino boards myself, and I do not know of anyone testing the library for that platform.

> If possible, what would be the advantages of a Teensy board over a Xiao?

Teensy boards use Cortex M4 and M7 cores, while the SAMD21 has a meager M0+, these are in completely different ballparks in terms of CPU performance. Teensies also have more RAM and flash memory, more advanced peripherals etc.  
It depends on what you want to do with it of course, for a simple MIDI controller without displays or audio, CPU performance is not going to be the bottleneck, the SAMD21 will be plenty fast and will probably have enough memory.

One particular difference is that Teensy boards support multiple virtual cables when using MIDI over USB, which is not possible when using the Arduino MIDIUSB library. 

Given the low price of the Seeeduino XIAO, it might be worth trying it out first, you can always upgrade to something more powerful later if you need to.
